### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by ecmendenhall**
This PR is a hack to get around the Github commenting interface. By opening a PR into my own fork, I can comment on all the latest code at once, instead of commit by commit.

### CODE REVIEW

File: Test/com/nicoles_change_maker/test/ChangeMakerTest.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
+package com.nicoles_change_maker.test;
+import com.nicoles_change_maker.ChangeMaker;
+import org.junit.*;
+import java.util.HashMap;
+
+import static org.junit.Assert.assertEquals;
+
+/**
```
Comment (ecmendenhall) [inline Test/com/nicoles_change_maker/test/ChangeMakerTest.java]: Here's a classic example of a garbage IDE comment. Everything included in this comment is already there in Git. The quickest way to remove these is to put your cursor over these lines in IntelliJ, hit Option + Enter, and choose "Edit Template." Then delete everything in the template text area and save. It should automatically remove all of these.

File: Test/com/nicoles_change_maker/test/ChangeMakerTest.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
+package com.nicoles_change_maker.test;
+import com.nicoles_change_maker.ChangeMaker;
+import org.junit.*;
```
Comment (ecmendenhall) [inline Test/com/nicoles_change_maker/test/ChangeMakerTest.java]: What are you actually using from JUnit? (I think it's actually just one class and one static method). It's a better practice to import each dependency explicitly by name: it gives readers a much clearer idea of how many dependencies this class actually has.

File: Test/com/nicoles_change_maker/test/ChangeMakerTest.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Created by bears8yourface on 1/5/16.
+ */
+
+public class ChangeMakerTest {
+    ChangeMaker changeMaker;
+    HashMap expectedChange = new HashMap<String,Integer>();
```
Comment (ecmendenhall) [inline Test/com/nicoles_change_maker/test/ChangeMakerTest.java]: I get compiler warnings on this line and every line below that uses this `HashMap`. Hover over the yellow bars in the scrollbar to read them:

generics

I know this code was from your First Day of Java Ever, but you're probably ready to read up on Generics, which are the source of this warning: https://docs.oracle.com/javase/tutorial/java/generics/index.html. Chapter 16 in Head First Java covers Generics, too.

The fix here is moving your type parameters to the other side of the equals sign:

``` java
HashMap<String, Integer> expectedChange = new HashMap<>();
```

...but I'll let you read the docs to understand why.

File: src/com/nicoles_change_maker/ChangeMaker.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Iterator;
+/**
+ * Created by bears8yourface on 1/5/16.
+ */
+public class ChangeMaker {
+    double remainingDollars;
+
+    public HashMap makeChange(double dollarAmount) {
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/ChangeMaker.java]: This method signature is also the source of a compiler warning, because the return type doesn't have type parameters. The Java compiler doesn't just want to know that this returns a `HashMap`, but what types of keys and values are inside. Is this a `HashMap<String, Integer>`, a `HashMap<Coin, Amount>`, a `HashMap<HashMap<String, String>,List<Integer>>`, or something even crazier? Your code will still compile without these annotations, but it's generally a good idea to include them.

File: Test/com/nicoles_change_maker/test/ChangeMakerTest.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    @Test
+    public void testChangeForEightCents(){
+        expectedChange.put("N", 1);
+        expectedChange.put("P", 3);
+        change = changeMaker.makeChange(.08);
+        assertEquals(expectedChange, change);
+    }
+
+    @Test
+    public void testChangeForNineCents(){
```
Comment (ecmendenhall) [inline Test/com/nicoles_change_maker/test/ChangeMakerTest.java]: Were these tests (one nickel, many pennies) driving your production code, or just verifying that it worked as you expected?

File: Test/com/nicoles_change_maker/test/ChangeMakerTest.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    public void testChangeForSeventeenCents(){
+        expectedChange.put("D", 1);
+        expectedChange.put("N", 1);
+        expectedChange.put("P", 2);
+        change = changeMaker.makeChange(.17);
+        assertEquals(expectedChange, change);
+    }
+
+    @Test
+    public void testChangeForEighteenCents(){
```
Comment (ecmendenhall) [inline Test/com/nicoles_change_maker/test/ChangeMakerTest.java]: Same question as above (https://github.com/ecmendenhall/ChangeMaker/pull/1/files#r50032542) down here. 

When I started testing, I thought writing unit tests was mostly about proving correctness, but I now realize that that's not their primary purpose. In most domains we can't possibly hope to test every single input and output. Looking at these, I suspect that they are some combination of:

1) Tests that are the "red" step in a red-green-refactor cycle, which enabled you to write new production code.
2) Tests that you wrote to verify your understanding, either of Java itself or of your own code. 
3) Tests that you wrote because I told you to TDD and it seemed like you should cover everything.

Keep in mind that tests are also code, and also rot over time without attention and maintenance. Tests of type 1) are very valuable: they are likely to cover many of the edge cases in whatever problem you're trying to solve. Tests of type 2) are sometimes useful, especially if other people are likely to have the same questions you did as you wrote this code. But you should think carefully before checking in tests of type 3). Are they actually necessary, or will they just rot over time? 

The red-green-refactor cycle includes refactoring tests, too! Eventually you'll develop a good feeling for paring test suites down to a minimum number of useful cases.

Comment (ecmendenhall) [inline Test/com/nicoles_change_maker/test/ChangeMakerTest.java]: Also, this doesn't mean you shouldn't write these kind of tests along the way, especially if they're helpful stepping stones! Just that they might not always need to be permanent.

File: Test/com/nicoles_change_maker/test/CoinMapTest.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.HashMap;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Created by bears8yourface on 1/5/16.
+ */
+public class CoinMapTest {
+    CoinMap coinMap = new CoinMap();
+    HashMap dictionary = coinMap.dictionary;
+    private static final double DELTA = .001;
```
Comment (ecmendenhall) [inline Test/com/nicoles_change_maker/test/CoinMapTest.java]: Is this used? The compiler is probably raising a warning if you check in your IDE.

File: Test/com/nicoles_change_maker/test/CoinMapTest.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Test
+    public void testPennyValue() {
+        assertEquals(.01, dictionary.get("P"));
```
Comment (ecmendenhall) [inline Test/com/nicoles_change_maker/test/CoinMapTest.java]: Here's an example of an object violating encapsulation and exposing its internals. Instead of calling a method on `CoinMap`, e.g.: 

``` java
assertEquals(.01, CoinMap.valueForCoin("P"));
```

I have to reach inside the object and pull out `dictionary`, a data structure. This class would be better encapsulated if it hid the `HashMap` internally, exposing the behavior I want ("tell me what the value of this coin is") without exposing the implementation ("look up the value of this coin in a `HashMap`)".

File: src/com/nicoles_change_maker/ChangeMaker.java (hunk 4 above)
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/ChangeMaker.java]: Floats and doubles are imprecise, and it's a good practice not to use them to represent money: https://stackoverflow.com/questions/3730019/why-not-use-double-or-float-to-represent-currency/3730249#3730249. 

(Same thing in Ruby: use `Integer` or `BigDecimal` instead).

File: src/com/nicoles_change_maker/ChangeMaker.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        while(coinIterator.hasNext()) {
+            Map.Entry coinValueMapping = (Map.Entry) coinIterator.next();
+            evaluateCoinAmountForInclusionInChangeCollection(changeCollection, coinValueMapping);
+        }
+
+        return changeCollection;
+    }
+
+    private void evaluateCoinAmountForInclusionInChangeCollection(HashMap changeCollection, Map.Entry coinValueMapping) {
+        String coinType = (String) coinValueMapping.getKey();
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/ChangeMaker.java]: Here's another spot where adding type parameters to generics comes in handy: You shouldn't need to cast these values to `String` and `Double`, because the compiler will already be sure that `coinValueMapping` keys are strings and values are doubles.

File: src/com/nicoles_change_maker/CoinMap.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        {
+            put("H", .50);
+            put("Q", .25);
+            put("D", .10);
+            put("N", .05);
+            put("P", .01);
+        }
+    };
+
+    public Iterator getIterator() {
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/CoinMap.java]: In typed languages, encoding types in names is a smell: it duplicates information that's already encoded in the type! I know that this method returns an `Iterator`, so naming it `getIterator` doesn't add much. What is it actually doing? (That is, what does this thing iterate over?)

File: src/com/nicoles_change_maker/CoinMap.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
+package com.nicoles_change_maker;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+/**
+ * Created by bears8yourface on 1/5/16.
+ */
+public class CoinMap {
+
+    public static final LinkedHashMap dictionary = new LinkedHashMap<String, Double>() {
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/CoinMap.java]: I mentioned this before, but you should more or less never expose a public instance variable. You have a bit more protection here because it's `final`, but it's always best to expose behavior, not data.

File: src/com/nicoles_change_maker/ChangeMaker.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        if (remainingDollars >= coinValue) {
+            int numOfCoins = howManyCoins(coinValue);
+            changeCollection.put(coinType, numOfCoins);
+            remainingDollars = calculateRemainingDollars(numOfCoins * coinValue);
+        }
+    }
+
+    private double calculateRemainingDollars(double amountCovered) {
+        double remainingDollarsExact = remainingDollars - amountCovered;
+        return Math.round(remainingDollarsExact * 100.0) / 100.0;
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/ChangeMaker.java]: :scream_cat: Here's a place where your `double`s could go wrong.

File: src/com/nicoles_change_maker/ChangeMaker.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    public HashMap makeChange(double dollarAmount) {
+        remainingDollars = dollarAmount;
+
+        if (checkDollarAmountIsValid(dollarAmount)) {
+            return evaluateEachCoinAmount();
+        }
+        throw new IllegalArgumentException();
+    }
+
+    private HashMap evaluateEachCoinAmount() {
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/ChangeMaker.java]: "Evaluate" doesn't seem like quite the right verb here. What does it actually mean to "evaluate a coin amount?"

File: src/com/nicoles_change_maker/ChangeMaker.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        HashMap changeCollection = new HashMap();
+        Iterator coinIterator = new CoinMap().getIterator();
+
+        while(coinIterator.hasNext()) {
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/ChangeMaker.java]: Modern Java has much nicer iteration abstractions, like the "enhanced for loop" (https://docs.oracle.com/javase/tutorial/java/nutsandbolts/for.html), which lets you avoid C-style `int i=0; i < collection.length(); i++` nonsense:

``` java
for (Map.Entry coinValueMapping : coinIterator) {
  // do stuff with coinValueMapping
}
```

In this context, using an iterator seems a little off, since you have to call `.hasNext()` and `next()`, which are really implementation details. If you go with an enhanced for loop, you can just use an _`Iterable`_ (https://docs.oracle.com/javase/7/docs/api/java/lang/Iterable.html), like a `LinkedList` or `ArrayList`, e.g.:

``` java
for (Entry valueMapping : coinMap.coinValues()) {
  evaluateCoinAmountForInclusionInChangeCollection(changeCollection, valueMapping);
}
```

File: src/com/nicoles_change_maker/ChangeMaker.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+            Map.Entry coinValueMapping = (Map.Entry) coinIterator.next();
+            evaluateCoinAmountForInclusionInChangeCollection(changeCollection, coinValueMapping);
+        }
+
+        return changeCollection;
+    }
+
+    private void evaluateCoinAmountForInclusionInChangeCollection(HashMap changeCollection, Map.Entry coinValueMapping) {
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/ChangeMaker.java]: :+1: for being unafraid of long names. But I'm not sure the length helps reveal the intention. Again, what does "evaluating" a coin mean, and what does it mean to "include a coin in a change collection". Plain language names are OK even if they sound goofy.

File: src/com/nicoles_change_maker/ChangeMaker.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        String coinType = (String) coinValueMapping.getKey();
+        Double coinValue = (Double) coinValueMapping.getValue();
+
+        handlePossibleCoinAddition(coinType, coinValue, changeCollection);
+    }
+
+    private void handlePossibleCoinAddition(String coinType, double coinValue, HashMap changeCollection) {
```
Comment (ecmendenhall) [inline src/com/nicoles_change_maker/ChangeMaker.java]: "Handle" is another weak verb. What's actually being handled?
